File: Pythonleetcode/Leetcode173.py
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.res=[]
        self.i=-1
        self.helper(root)


    def next(self):
        """
        :rtype: int
        """
        self.i+=1
        return self.res[self.i]

    def hasNext(self):
        """
        :rtype: bool
        """
        if self.i<len(self.res)-1:
            return True
        else:
            return False
    def helper(self,root):
        if not root:
            return
        self.helper(root.left)
        self.res.append(root.val)
        self.helper(root.right)

File: CppPractice/CppLeetcode/Leetcode173.cpp
#include<vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode():val(0),left(nullptr),right(nullptr){};
    TreeNode(int x):val(x),left(nullptr),right(nullptr){};
    TreeNode(int x, TreeNode *leftp,TreeNode *rightp):val(x),left(leftp),right(rightp){};
};

class BSTIterator {
private:
    vector<TreeNode*> theList;
    void PushLeft(TreeNode *p)
    {
        if (p)
        {
            theList.push_back(p);
            PushLeft(p->left);
        }
        return;
    }
public:
    BSTIterator(TreeNode* root) {
        PushLeft(root);
        
    }
    
    int next() 
    {
        TreeNode *nextP = theList.back();
        theList.pop_back();
        PushLeft(nextP->right);
        return nextP->val;
        
    }
    
    bool hasNext() 
    {
        return (theList.size()>0);
        
    }
};
